integration/tests/epochs/epoch_test.go

```diff
@@ -52,36 +62,79 @@ func (s *Suite) TestViewsProgress() {
 		epochFinal, err := epoch.FinalView()
 		require.NoError(s.T(), err)
 
-		expectedViews := []testCase{
-			{epoch: epochCounter, view: epochFirstView, phase: flow.EpochPhaseStaking},
-			{epoch: epochCounter, view: epochDKGPhase1Final, phase: flow.EpochPhaseSetup},
-			{epoch: epochCounter, view: epochDKGPhase2Final, phase: flow.EpochPhaseSetup},
-			{epoch: epochCounter, view: epochDKGPhase3Final, phase: flow.EpochPhaseSetup},
-			{epoch: epochCounter, view: epochFinal, phase: flow.EpochPhaseCommitted},
+		epochViews := []*phaseCheck{
+			{epoch: epochCounter, phase: flow.EpochPhaseStaking, finalView: epochFirstView},
+			{epoch: epochCounter, phase: flow.EpochPhaseSetup, finalView: epochDKGPhase1Final},
+			{epoch: epochCounter, phase: flow.EpochPhaseSetup, finalView: epochDKGPhase2Final},
+			{epoch: epochCounter, phase: flow.EpochPhaseSetup, finalView: epochDKGPhase3Final},
+			{epoch: epochCounter, phase: flow.EpochPhaseCommitted, finalView: epochFinal},
```

Comment:
⚠️  This is still very susceptible to races. We are asserting that the latest sealed block with view >= `epoch.FinalView` should be within the epoch. The only way this is true is if we:
* finalize a block with exactly `view==epoch.FinalView` since any blocks with greater views will be in the *next* epoch (there will be at most one such block, and it is possible no such block will be proposed or finalized) and
* capture this block in our query to `WaitForSealedView` (also not guaranteed because this query is greedy and not synchronized with the rest of the `BlockState` data structure)

In general this assertion does not hold, even without a race condition in our query. In running these tests, I have seen both (1) the `epochFinal` phase check fail (this is expected given the assertion we are making) and (2) the `epochFirstView` phase check fail (this is unexpected and I think reflects a race or bug in the `WaitForSealedView` query).

### Issue 1
To address (1) we need to change the assertion. In particular, I would suggest we query for the latest sealed block with `view<=epoch.FinalView`. This block is guaranteed to exist and to be within the `EpochCommit` phase if the epoch transition was successful. Here's an example of a test failure I observed in this category:
```
            	Test:       	TestEpochs/TestViewsProgress
            	Messages:   	wrong phase at view 179
        epoch_test.go:104:
            	Error Trace:	epoch_test.go:104
            	Error:      	Not equal:
            	            	expected: 0x0
            	            	actual  : 0x1
            	Test:       	TestEpochs/TestViewsProgress
            	Messages:   	wrong epoch at view 179 (block: height=177, view=180)
        epoch_test.go:108:
            	Error Trace:	epoch_test.go:108
            	Error:      	Not equal:
            	            	expected: 3
            	            	actual  : 1
            	Test:       	TestEpochs/TestViewsProgress
            	Messages:   	wrong phase at view 179 (block: height=177, view=180)
```
The final view is 179. We instead receive the block with view 180 (in the next epoch), either because no block exists with view 179 or because of the race condition in the query.


### Issue 2
To address (2) we need to fix the race condition. The test failure I observed in this category is:
```
        epoch_test.go:104:
            	Error Trace:	epoch_test.go:104
            	Error:      	Not equal:
            	            	expected: 0x1
            	            	actual  : 0x0
            	Test:       	TestEpochs/TestViewsProgress
            	Messages:   	wrong epoch at view 180 (block: height=178, view=179)
        epoch_test.go:108:
            	Error Trace:	epoch_test.go:108
            	Error:      	Not equal:
            	            	expected: 1
            	            	actual  : 3
            	Test:       	TestEpochs/TestViewsProgress
            	Messages:   	wrong phase at view 180 (block: height=178, view=179)
```

Here we are attempting to check the first view (180) but the query returns us a block with view 179, indicating a bug in the `WaitForSealedView`. I'm not sure of the cause, but here are a couple of ideas:
* The `WaitForSealedView` method is not synchronized with `Add` so there can be concurrent reads and writes to shared `BlockState` variables
* A variety of factors can cause blocks to be added out of order and sealed in quick succession, so that a polling approach may miss the exact block we are querying for. Instead, we could wait for _a_ block with `view>=queryView`, then walk down the `finalizedByHeight` state by height to find the exact block which optimally matches the query (the lowest `view` such that `view>=queryView`). I suspect this approach would be desirable for addressing (1) as well. 

**Final note:** Yahya has implemented some fixes to the `BlockState` type to address some flakey tests in https://github.com/onflow/flow-go/pull/936. I've merged those changes into `feature/epochs`, but please merge them into this branch as well before continuing here (they might help with some of these issues and at least will avoid conflicts when we are merging this with `master`).

Reply:
For issue 2, it's because I am taking the **Parent** of the block that is returned by WaitForSealed:
https://github.com/onflow/flow-go/blob/7605bed7b304767d10dce7424863be209e51b72a/integration/tests/epochs/epoch_test.go#L75

There is an edge case at the beginning and end of an epoch I think

I take the parent because, as you noted, the block returned by WaitForSealed is not necessarily in the epoch. And I thought that the Parent would be the latest sealed block where view <= TargetView... but I might need to look into WaitForSealed more closely

Unfortunately the tests are passing on my machine so I can't reproduce your errors..

Reply:
The problem comes down to finding a reliable way to get the last finalized block with a view <= a target view.
Even using the parent of block returned by WaitForSealed doesn't work.
BlockState needs to be reworked.. or we need to find a way without BlockState

Reply:
Ok, I took a different approach to this that doesn't use `BlockState`.
Once the containers are stopped, I build a map of view to { epoch, phase } by inspecting the state.
Then I check this map against what is expected
https://github.com/onflow/flow-go/blob/3e5a008b93e886576db0675f39aa4a54f77306a1/integration/tests/epochs/epoch_test.go#L99-L118